(TK-159) Round Two of ConfigValue tests

jpinsonault opened the pull request:
Another set of tests for config_value_spec. There will be at least one more after this one.

Added a few methods to AbstractConfigObject
Made the definitions of merge_origins static to match the java
Implemented render/render_to_sb methods for ConfigDelayedMerge

Fixed DefaultTransformer::transform method to actually compare the value type

Made SimpleConfigObject::indent static

SimpleConfigOrigin::merge_origins handles merging more than two tokens correctly

Implemented SimpleConfigOrigin#filename using @cprice404's Url class

A couple methods in tokens.rb
  get_substitution_path_expression
  get_substitution_optional

Commented out some failing tests in conf_parser_spec until some missing functionality is implmented

cprice404 commented on lib/hocon/impl/abstract_config_object.rb:
```diff
@@ -21,6 +21,47 @@ def initialize(origin)
+      object
+    end
+  end
+
```
@jpinsonault is this PR based on top of the one we merged yesterday-ish where I'd taken a pass through AbstractConfigObject?  I thought I'd gotten that class pretty much synced up, so I'm wondering how these were missing?

fpringvaldsen replied:
Also, I don't think these methods should be implemented in `AbstractConfigObject`, as these are all supposed to be `abstract` methods.

jpinsonault replied:
It definitely is. That's... a very good question, I'm confused

jpinsonault replied:
@fpringvaldsen Ah good point

fpringvaldsen commented on lib/hocon/impl/config_delayed_merge.rb:
```diff
@@ -56,4 +56,78 @@ def hash
+      if comment_merge
+        self.indent(sb, indent, options)
+        if !at_key.nil?
+          sb << "#     unmerged value #{i} for key "
```
Looks like there should be a call to `ConfigImplUtil.renderJsonString` here

jpinsonault replied:
Good catch

fpringvaldsen commented on spec/test_utils.rb:
```diff
@@ -119,8 +119,12 @@ def whitespace_matters?
       ParseTest.new(false, true, "[${ ?foo}]"), # space before ? not allowed
-      ParseTest.from_s(%q|{ "a" : [1,2], "b" : y${a}z }|), # trying to interpolate an array in a string
-      ParseTest.from_s(%q|{ "a" : { "c" : 2 }, "b" : y${a}z }|), # trying to interpolate an object in a string
+      # Commenting out the following 2 tests because we suspect missing code in SimpleConfigList/Object
```
Might want to add a TODO here so we remember to come back and fix it later

fpringvaldsen commented on spec/unit/typesafe/config/config_value_spec.rb:
```diff
@@ -506,3 +533,115 @@
+    empty = TestUtils.parse_config("{}")
+
+    expect(empty.has_path("foo")).to be_falsey
+    expect(empty.has_path("foo")).to be_falsey
```
I think we should only need one of these

jpinsonault replied:
Apparently I REALLY wanted to make sure it was false

fpringvaldsen commented:
Looking good! I just had a few minor comments.

jpinsonault commented:
@fpringvaldsen Added commit to address your feedback

* Removed redefinitions of the abstract methods
* Added call to render_json_string
* Added TODO comment
* Removed my over enthusiastic false checks

cprice404 commented:
I took a quick pass over this, looks good so far; I'm worried about SimpleConfigOrigin though so I think we should merge that other PR and rebase this and then I'll review it in more detail.

jpinsonault commented:
@cprice404 agreed

fpringvaldsen commented:
:+1:

jpinsonault commented:
@fpringvaldsen @cprice404 Just rebased on @cprice404's PR

cprice404 commented on lib/hocon/impl/config_delayed_merge.rb:
```diff
@@ -56,4 +56,79 @@ def hash
     @stack.hash
   end
+
+  def render_to_sb(sb, indent, at_root, at_key, options)
```
So, I may just be confused about the ways we've had to rename the various render methods to deal with all of the signature overloads, but I was thinking that the one that you need to override in subclasses of AbstractConfigObject was `render_value_to_sb`.  If this is working for you then I probably just have some assumption wrong somewhere, but do you know where this gets called from?  (I'll poke through the code to look for that in a sec, after I've reviewed the rest of the PR.)

jpinsonault replied:
My brain hurts remembering doing this, but I think that I used `render_to_sb` because I needed the `at_key` parameter to call `render_value_to_sb_from_stack`...

cprice404 replied:
ah, yep, you're right.

cprice404 commented on lib/hocon/impl/config_delayed_merge_object.rb:
```diff
@@ -170,4 +170,8 @@ def hash
     @stack.hash
   end
+
+  def render_to_sb(sb, indent, at_root, at_key, options)
```
Same question here; not sure if this should be `render_value_to_sb`.  Looking at the code, I *think* the `render_to_sb` in `AbstractConfigValue` is supposed to be generic and re-used across all subclasses, and then at the end of it it calls `render_value_to_sb` which is the one that should be specific per subclass... but we probably need to take a quick look at the upstream signatures and validate that.

jpinsonault replied:
I get very confused when I look at this, but I think where I came from was I looked at `protected void render(StringBuilder sb, int indent, boolean atRoot, String atKey, ConfigRenderOptions options)` in `ConfigDelayedMerge`, saw the signature,  and found the corresponding ruby method `render_to_sb(sb, indent, at_root, at_key, options)`

cprice404 commented:
+1 except for making sure that we're overriding the correct `render` methods.
